common/verityfs.c

```diff
@@ -0,0 +1,340 @@
+		*p++ = hex[tmp & 15];
+	}
+	*p = '\0';
+}
+
+static int
+generate_dm_table_load_extra_params(uint8_t *buf, size_t len, verity_sb_t *sb, char *fs_dev,
+				    uint64_t fs_size, char *hash_dev, const char *root_hash)
+{
+	if (len < sizeof(struct dm_ioctl) + sizeof(struct dm_target_spec)) {
+		ERROR("Failed to generate dm_table_load extra params: Buffer too small");
+		return -1;
+	}
+
+	struct dm_target_spec *tgt;
+	tgt = (struct dm_target_spec *)&buf[sizeof(struct dm_ioctl)];
+	tgt->sector_start = 0;
+	tgt->length = fs_size;
+	tgt->status = 0;
+	strncpy(tgt->target_type, "verity", sizeof(tgt->target_type));
+
+	char *verity_params;
+	verity_params = (char *)(buf + sizeof(struct dm_ioctl) + sizeof(struct dm_target_spec));
+
+	char *salt = convert_bin_to_hex_new(sb->salt, sb->salt_size);
+	uint32_t offset = 1;
+	snprintf(verity_params, len - sizeof(struct dm_ioctl) - sizeof(struct dm_target_spec),
+		 "%d %s %s %d %d %ld %d %s %s %s", sb->version, fs_dev, hash_dev,
+		 sb->data_block_size, sb->hash_block_size, sb->data_blocks, offset, sb->algorithm,
+		 root_hash, salt);
+
+	mem_free(salt);
+
+	// Set pointer behind parameter
+	verity_params += strlen(verity_params) + 1;
+	// Align to an 8 byte boundary
+	verity_params = (char *)(((unsigned long)verity_params + 7) & ~8);
+	// Set tgt->next right behind dm_target_spec
+	tgt->next = (unsigned int)(verity_params - (char *)buf);
+
+	return 0;
+}
+
+int
+verityfs_open(const char *name, const char *fs_img_name, const char *hash_dev_name,
+	      const char *root_hash)
+{
+	int control_fd = -1;
+	int ret = -1;
+	uint8_t buf[16384] = { 0 };
+	struct dm_ioctl *dmi = NULL;
+
+	TRACE("Opening dm-verity device %s with image %s, hash-tree %s and root-hash %s", name,
+	      fs_img_name, hash_dev_name, root_hash);
+
+	// Get UUID and Superblock information from hash device
+	verity_sb_t sb = { 0 };
+	int hash_fd = open(hash_dev_name, O_RDONLY);
+	if (hash_fd < 0) {
+		ERROR_ERRNO("Failed to open dm-verity hash device %s", hash_dev_name);
+	}
+
+	ssize_t size = fd_read_blockwise(hash_fd, &sb, sizeof(verity_sb_t), 4096, 4096);
+	if (size < (ssize_t)sizeof(verity_sb_t)) {
+		ERROR("Failed to read superblock of %s", hash_dev_name);
+		goto out;
+	}
+	close(hash_fd);
+
+	// Create dm-verity format uuid
+	char uuid[40] = { 0 };
+	char dev_uuid[DM_UUID_LEN] = { 0 };
+	uuid_bytes_to_string(uuid, sb.uuid);
+	if (verity_create_uuid(name, uuid, dev_uuid, sizeof(dev_uuid))) {
+		ERROR("Failed to create uuid for dm-verity device %s", name);
+		goto out;
+	}
+	TRACE("dm-verity device %s uuid: %s\n", name, dev_uuid);
+
+	control_fd = open(DM_CONTROL, O_RDWR);
+	if (control_fd < 0) {
+		ERROR_ERRNO("Failed to open %s\n", DM_CONTROL);
+		goto out;
+	}
+
+	// Make sure that dm-verity device does not already exist
+	dmi = (struct dm_ioctl *)buf;
+	dm_ioctl_init(dmi, INDEX_DM_TABLE_STATUS, sizeof(buf), name, NULL, DM_EXISTS_FLAG, 0, 0, 0);
+	dm_ioctl(control_fd, cmd_table[INDEX_DM_TABLE_STATUS].cmd, dmi);
+	if (errno != ENXIO) {
```

Comment:
should we not also check the return value of dm_ioctl !=0 ? Is this intentionally left out?

Reply:
Left out, as an error is expected here. But of course, we can check for ret == 0, I added this